server/repository/src/db_diesel/changelog/changelog.rs

```diff
@@ -205,16 +378,80 @@ fn create_filtered_query(earliest: u64, filter: Option<ChangelogFilter>) -> Boxe
             action,
+            source_site_id,
         } = f;
 
         apply_equal_filter!(query, table_name, changelog_deduped::table_name);
         apply_equal_filter!(query, name_id, name_link::name_id);
         apply_equal_filter!(query, store_id, changelog_deduped::store_id);
         apply_equal_filter!(query, record_id, changelog_deduped::record_id);
-        apply_equal_filter!(query, is_sync_update, changelog_deduped::is_sync_update);
         apply_equal_filter!(query, action, changelog_deduped::row_action);
+        apply_equal_filter!(query, is_sync_update, changelog_deduped::is_sync_update);
+        apply_equal_filter!(query, source_site_id, changelog_deduped::source_site_id);
+    }
+
+    query
+}
+
+/// This looks up associated records to decide if change log should be sent to the site or not
+/// Update this method when adding new sync styles to the system
+fn create_filtered_outgoing_sync_query(
+    earliest: u64,
+    sync_site_id: i32,
+    is_initialized: bool,
+) -> BoxedChangelogQuery {
+    let mut query = changelog_deduped::table
+        .left_join(name_link::table)
+        .filter(changelog_deduped::cursor.ge(earliest.try_into().unwrap_or(0)))
+        .into_boxed();
+
+    // If we are initialising, we want to send all the records for the site, even ones that originally came from the site
+    // The rest of the time we want to exclude any records that were created by the site
+
+    if is_initialized {
+        query = query
+            .filter(
+                changelog_deduped::source_site_id
+                    .ne(Some(sync_site_id.clone()))
+                    .or(changelog_deduped::source_site_id.is_null()),
+            )
+            .filter(changelog_deduped::is_sync_update.eq(false))
     }
 
+    // Loop through all the Sync tables and add them to the query if they have the right sync style
+
+    // Central Records
+
+    let central_sync_table_names: Vec<ChangelogTableName> = ChangelogTableName::iter()
+        .filter(|table| match table.sync_style() {
+            ChangeLogSyncStyle::Central => true,
+            _ => false,
+        })
+        .collect();
+
+    // Remote Records
+    let remote_sync_table_names: Vec<ChangelogTableName> = ChangelogTableName::iter()
+        .filter(|table| match table.sync_style() {
+            ChangeLogSyncStyle::Remote => true,
+            _ => false,
+        })
+        .collect();
+
+    let active_stores_for_site = store::table
+        .filter(store::site_id.eq(sync_site_id))
+        .select(store::id.nullable())
+        .into_boxed();
+
+    // Filter the query for the matching records for each type
+    query = query.filter(
+        changelog_deduped::table_name
+            .eq_any(central_sync_table_names)
+            .or(changelog_deduped::table_name
+                .eq_any(remote_sync_table_names)
+                .and(changelog_deduped::store_id.eq_any(active_stores_for_site))), // TODO Should this use name_link?
```

Comment:
Yeah It might be better with raw sql. Was thinking this does give us the count vs results query in sync and does allow some level of type checking etc.